mlflow/pyfunc/__init__.py

```diff
@@ -266,7 +270,139 @@ def _load_model_env(path):
 
 
-def load_model(model_uri, suppress_warnings=True):
+def _enforce_type(name, values: pandas.Series, t: DataType):
+    """
+    Enforce the input column type matches the declared in model input schema.
+
+    The following type conversions are allowed:
+
+    1. np.object -> string
+    2. int -> long
+    3. float -> double.
+
+    Any other type mismatch will raise error.
+    """
+    if values.dtype == np.object and t not in (DataType.binary, DataType.string):
+        values = values.infer_objects()
+
+    if values.dtype in (t.to_pandas(), t.to_numpy()):
+        # The types are already compatible => conversion is not necessary.
+        return values
+
+    if t == DataType.binary and values.dtype.kind == t.binary.to_numpy().kind:
+        #  NB: bytes in numpy have variable itemsize and would fail the upcast check.
+        return values
+
+    if t == DataType.string and values.dtype == np.object:
+        #  NB: strings are by default parsed and inferred as objects, but it is
+        # recommended to use StringDtype extension type if available. See
+        #
+        # `https://pandas.pydata.org/pandas-docs/stable/user_guide/text.html`
+        #
+        # for more detail.
+        try:
+            return values.astype(t.to_pandas(), errors="raise")
+        except ValueError:
+            raise MlflowException(
+                "Failed to convert column {0} from type {1} to {2}.".format(
+                  name, values.dtype, t)
+            )
+
+    numpy_type = t.to_numpy()
+    is_compatible_type = values.dtype.kind == numpy_type.kind
+    is_upcast = values.dtype.itemsize <= numpy_type.itemsize
+    if is_compatible_type and is_upcast:
+        # upcasting compatible types is ok
+        return values.astype(numpy_type, errors="raise")
+    else:
+        # NB: conversion between incompatible types (e.g. floats -> ints or
+        # boolean -> numeric) are not allowed. While supported by pandas and numpy,
+        # these conversions alter the values significantly.
+        raise MlflowException("Incompatible input types for column {0}. "
+                              "Can not safely convert {1} to {2}.".format(name,
+                                                                          values.dtype,
+                                                                          numpy_type))
+
+
+def _enforce_schema(pdf: pandas.DataFrame, input_schema: Schema):
+    """
+    Enforce column names and types match the input schema.
+
+    For column names, we check there are no missing columns and reorder the columns to match the
+    ordering declared in schema if necessary. Any extra columns are ignored.
+
+    For column types, we make sure the types match schema or can be safely converted to match the
+    input schema.
+    """
+    col_names = input_schema.column_names()
+    col_types = input_schema.column_types()
+    expected_names = set(col_names)
+    actual_names = set(pdf.columns)
```

Comment:
Just to confirm, we have a test that ensures that this works for unnamed columns?

Reply:
Good catch, we don't! In fact, now that I am thinking about it, it probably makes more sense to not check column names at all if schema did not declare them. We should only check that we got at least the expected number of columns. We actually enforce them to be an integer range which is weird. I'll update the logic and add a test.